jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipResponseAndCallback.java

```diff
@@ -220,14 +222,23 @@ protected void commit(boolean last, Callback callback, ByteBuffer content)
             return;
         }
 
+        // We are varying the response due to accept encoding header.
+        if (_vary != null)
+            fields.ensureField(_vary);
```

Comment:
What is the reason to move this here?
And why not put it before the other exclusions?

Reply:
From [RFC 9110](https://www.rfc-editor.org/rfc/rfc9110.html#field.vary):

> The "Vary" header field in a response describes what parts of a request message, aside from the method and target URI, might have influenced the origin server's process for selecting the content of this response.

So in theory we should always add it when we have the GzipHandler installed, as any response might change since it may be compressed.

However, this will break a number of tests.

Not sure if we interpreted the semantic wrong before (and wrote wrong tests) but I tried to make the minimum amount of changes in this PR.

However, I do think we should square this out once and for all.

@lachlan-roberts your thoughts?

Reply:
@lachlan-roberts not sure if the interpretation could be:

If you ask the server for an empty response, it will always be empty and not compressed so it will be independent from the `Accept-Encoding` header, so it should not have the `Vary` header.

However, the RFC says "aside from the method and target URI".
With these 2 excluded, it basically becomes an announcement of the capabilities of the server, so it should always be there.

I'm not sure what exactly the purpose is, if the method and URI are not considered.

Reply:
> If you ask the server for an empty response, it will always be empty and not compressed so it will be independent from the Accept-Encoding header, so it should not have the Vary header.

I can see how that would make sense. But the response might not always be empty so I think its better to always advertise the servers capability to vary the response on the `Accept-Encoding` header. Should probably just set it in `GzipHandler` instead of `GzipResponseAndCallback`.

Also what about the case where we skip compression because the `CONTENT_ENCODING` has already been set? right now we expect someone else to set the vary header.